### PR TITLE
10161: watch for changing base during Mul flattening

### LINKED ISSUE #10161: factor gives an invalid expression
This was reported on the mailing list a while back

```
In [29]: h = 2*x*(-2*x + Abs(x))*(x**2 - 1)/Abs(x**2 - 1) + (x/Abs(x) - 2)*Abs(x**2 - 1)

In [30]: print(factor(h))
(x - 1)*(x**4 - 6*x**3*Abs(x) + x**3 - 6*x**2*Abs(x) + x**2 + 2*x*Abs(x) - x + 2*Abs(x))/(Abs(x)*Abs(x**2 - 1))

In [31]: h.subs(x, 2)
Out[31]: -11

In [32]: factor(h).subs(x, 2)
Out[32]: -53/3
```

### PR DESCRIPTION
**Pull request by smichr**
fixes #10161

### CODE REVIEW

**Comment by asmeurer** [comment]
All the Travis builds are erroring with "No output has been received in the last 10 minutes, this potentially indicates a stalled build or something wrong with the build itself." I'm guessing this means there is a nontrivial performance hit here.

**Comment by smichr** [comment]
> nontrivial performance hit

:-)  not really, just a garden-variety infinite loop that I have to watch out for

File: sympy/core/mul.py [hunk 1]
```diff
@@ -402,14 +402,30 @@ def _gather(c_powers):
...
+            ebases = []
+            changed = False
+            for p, e in c_powers:
+                if e.is_zero:
+                    continue
+                if e is S.One:
+                    if p.is_Number:
+                        coeff *= p
+                        continue
+                b = bi = p
```
Comment (smichr) [inline sympy/core/mul.py]: put this into `else` since `e == 1`

File: sympy/core/mul.py [hunk 2]
```diff
@@ -402,14 +402,30 @@ def _gather(c_powers):
...
+                p = Pow(p, e)
+                if p.is_Pow:
+                    b, e = p.as_base_exp()
+                    if b != bi:
```
Comment (smichr) [inline sympy/core/mul.py]: add 'and not bi.is_Pow` to avoid unevaluated powers (and hopefully fix the single failing test)

**Comment by smichr** [comment]
I think this is ready to look at if the tests pass.

**Comment by smichr** [comment]
@jksuom, I acknowleged you in the 2nd commit -- is that OK or would you like to be the author of that commit?

**Comment by jksuom** [comment]
That is (more than) OK, thank you. Besides, you found a way of improving the code :-)
(Only, the test would rather belong to ``test__parallel_dict_from_expr_no_gens``.)

**Comment by smichr** [comment]
Got the test correction, @jksuom . Does the rest looks non-invasive enough to core?

**Comment by jksuom** [comment]
I hope there would be tests where the part using ``as_base_exp`` were exercised. I find it hard to see in which cases the loop would have to be repeated and what would happen then to ``coeff``.

**Comment by smichr** [comment]
> I hope there would be tests where the part using as_base_exp were exercised

That is what the 10161 test is using. The loop needs to be re-run to gather the `x` and the new `x**2` (from `abs(x)**2`) from `x*abs(x)*abs(x)`. As for the coefficient being updated...that was code that was presumably needed if the loop runs once but I didn't check that for coverage; it probably won't be hit while running it the second time. I'll try to see how it's covered later today.

**Comment by jksuom** [comment]
I have taken some more time studying the code. It looks a bit intricate on the surface but that probably cannot be helped. Also, I now see that ``coeff`` will be updated only once as it is not appended to ``new_c_powers``.
On the whole, the modifications are quite conservative. Only one repetition is allowed. (That can be changed later if necessary, but I think that is improbable.) I don't think there would be noticeable effect on performance. I am +1 for this.

**Comment by smichr** [comment]
Thanks, @jksuom , for looking this over. I'll go ahead and commit it as is.

**Pull request merged**